salt_stack/salt/backup/files/vault.py

```diff
@@ -0,0 +1,186 @@
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# A script to backup and restore data from Vault
+# Currently backs up and restores
+# * KV data under secret/
+# * Vault policies
+# * AWS-EC2 Authentication configuration
+# * AWS secret backend configuration
+#
+# NOTE: Restore doesn't restore AWS credentials (the credentials Vault uses to
+#       communicate with AWS). Restore expects to restore into a Vault that has
+#       already been configured (using lib/vault.py:Vault.configure) so that
+#       the different backends are ready for data
+
+from bossutils.vault import Vault
+import sys
+import os
+import json
+
+def export(v, path):
+    """Recursive function for exporting all paths and data"""
+    # DP NOTE: Taken from lib/vault.py:Vault.export
+    if path[-1] != '/':
+        path += '/'
+
+    rtn = {}
+
+    results =  v.client.read(path[:-1])
+    if results is not None:
+        rtn[path[:-1]] = results['data']
+
+    results = v.client.read(path + "?list=true")
+    for key in results['data']['keys']:
+        key = path + key
+        if key[-1] == '/':
+            data = export(v, key)
+            rtn.update(data)
+        else:
+            data = v.client.read(key)
+            rtn[key] = data['data']
+
+    return rtn
+
+if __name__ == "__main__":
+    # usage (backup|restore) domain
+    a = sys.argv[1]
+    d = sys.argv[2]
+
+    with open("/etc/boss/boss.config", "w") as fh:
+        fh.write("""[system]
+type = backup
+
+[vault]
+url = http://vault.{}:8200
+token =
+""".format(d))
+
+    v = Vault()
+
+    if a == "backup":
+        f = os.path.join(os.environ['OUTPUT1_STAGING_DIR'], 'export.json')
+        data = {
+            'policies': {},
+            'secrets': {},
+            'aws-ec2': {},
+            'aws': {},
+        }
+
+        # Backup policies
+        for policy in v.client.list_policies():
+            data['policies'][policy] = v.client.read('/sys/policy/' + policy)['rules']
+
+        # Backup secrets
+        data['secrets'] = export(v, 'secret/')
+
+        # Backup AWS secret backend roles
+        # DP ???: are these now automatically generated for ingest jobs?
+        #         if so, should they even be backed up?
+        prefix = '/aws/roles'
+        for role in v.client.read(prefix + '?list=true')['data']['keys']:
+            data['aws'][role] = v.client.read(prefix + '/' + role)['data']['policy']
+
+        # Backup AWS-EC2 auth backend roles
+        prefix = '/auth/aws-ec2/role'
+        for role in v.client.read(prefix + '?list=true')['data']['keys']:
+            d = v.client.read(prefix + '/' + role)['data']
```

Comment:
This overwrites the `d` defined at the top, right (argv[2]). Is that `d` even used?